### ISSUE TITLE
Failure on Backup- Win 10- Index was outside the bounds of the array error

### ISSUE THREAD

**Issue by JD4x4** [jezzab/DUMLdore#4]
1st time using DUMLdore, attempting Backup of .300 Spark on Win 10 OS. Failed 3 times with same "Index was outside the bounds of the array" error. 

Operation stops immediately after decrypting and displaying "Fixing headers", leaves form on screen if Continue is selected but is hung. Form clears if Quit is selected.

Creates backup & decrypt folders, populates with normal size .sig files in backup, zero byte .sig files in decrypt.
Does not create a bin file. Is it only supposed to create sig files??

1st & 2nd failures using local Git clone source files copied to 2 different folders/hds. 3rd try using dl'ed zip file from GitHub jezzab/DUMLdore master.

![backup error](https://user-images.githubusercontent.com/7014079/28602552-b38c00c4-718c-11e7-81d5-0531be0a2356.JPG)

`See the end of this message for details on invoking 
just-in-time (JIT) debugging instead of this dialog box.

************** Exception Text **************
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at WindowsFormsApplication1.Form1.btnBackupFW_Click(Object sender, EventArgs e)
   at System.Windows.Forms.Control.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnMouseUp(MouseEventArgs mevent)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ButtonBase.WndProc(Message& m)
   at System.Windows.Forms.Button.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.OnMessage(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)

**Comment by jezzab** [comment]
There seems to be an issue with OpenSSL and it being called (openssl.exe). I am struggling to replicate the issue.

**Comment by JD4x4** [comment]
ty. C'mon over to my house & use my PC if ya like. I've got some beers, but going out of town on Friday night, lol.

**Comment by JD4x4** [comment]
Just for giggles, I ran Win compatibility checker on openssl.exe and it suggested running it in Win 8 compatibility mode, so I set it to that & tried again. No change, same error.

You sure it's not some obscure internal variable for Windoze Forms that rarely gets used & isn't getting set? (I know nothing.. just throwing stuff out there).

**Comment by jezzab** [comment]
No the program downloads the sig files into /backup/ then runs openssl to decrypt them into /decrypt/ then reads them in and fixes each header. Its failing because openssl failed and the files in decrypt are 0 bytes. 
I have not been able to replicate it and no one has had this error

**Comment by JD4x4** [comment]
Well. That's just the story of my life. I checked openssl.exe in both my dl'ed zip and in my local clone, both the same 536 kb and the same CRC/SHA. 

I do have a Linux Mint box so I can try a backup with pyDUML. Is openssl used only when doing a backup, or is it needed for f/w loading also? I'd hate to be totally shackled to my Linux box, but if that's my only option ..

**Comment by jezzab** [comment]
Its only needed for making a backup.

**Comment by JD4x4** [comment]
I'm not seeing a way to easily make a f/w backup with pyDUML. Is there a way to finish the backup steps manually using DUMLdore & tools in folder? i.e. using a cmd window in the DUMLdore folder on the sig files it's already pulled from the AC?

**Comment by JD4x4** [comment]
Problem found, backup succeeded.
I opened a dos window in my working dir-
>D:\DUMLdore>openssl ?
This version of D:\DUMLdore\openssl.exe is not compatible with the version of Windows you're running. Check your computer's system information and then contact the software publisher.
>
>D:\DUMLdore>

Then found [https://wiki.openssl.org/index.php/Binaries](url)
where I found -

> Pre-compiled Win32/64 libraries without external dependencies to the Microsoft Visual Studio Runtime DLLs, except for the system provided msvcrt.dll.

with the link to [https://indy.fulgan.com/SSL/](url) which is a repo for compiled binaries for various OS's

Replaced openssl.exe, libeay32.dll, and ssleay32.dll with the files from there for my Win10 x86 system. You might want to distribute these instead, as they might be more universal.

Lastly, it would be great if there was an option (or switch?) to keep the decrypted sig folder & files in case anyone wants to experiment with them. This assuming that Assistant param changes might be reflected in them?
